File: src/2831_找出最长等值子数组.py
# 2831_找出最长等值子数组
# https://leetcode.cn/problems/find-the-longest-equal-subarray/

from typing import List


class Solution:
    def longestEqualSubarray(self, nums: List[int], k: int) -> int:
        n = len(nums)

        # Indexes list stores indices for each unique number in nums.
        indexes_list = [[] for _ in range(n + 1)]

        for idx, num in enumerate(nums):
            indexes_list[num].append(idx)

        max_equal_length = 0

        for indexes in indexes_list:
            if len(indexes):
                j = 0
                for i in range(len(indexes)):
                    # Find the farthest index j such that the number of gaps
                    # between i and j is not more than k.
                    while j < len(indexes) and indexes[j] - indexes[i] - (j - i) <= k:
                        j += 1
                    max_equal_length = max(max_equal_length, j - i)

        return max_equal_length

File: src/2831_找出最长等值子数组.cpp
#include <string>
#include <algorithm>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution
{
public:
    int longestEqualSubarray(vector<int> &nums, int k)
    {
        int n = nums.size();
        unordered_map<int, vector<int>> indexes_map;

        // Store indices for each number
        for (int i = 0; i < n; ++i)
        {
            indexes_map[nums[i]].push_back(i);
        }

        int max_equal_length = 0;

        for (auto &p : indexes_map)
        {
            vector<int> &indexes = p.second;
            int j = 0;
            for (int i = 0; i < indexes.size(); ++i)
            {
                // Expand j to maintain the condition
                while (j < indexes.size() && indexes[j] - indexes[i] - (j - i) <= k)
                {
                    j++;
                }
                max_equal_length = max(max_equal_length, j - i);
            }
        }

        return max_equal_length;
    }
};